Automattic/woocommerce-services
# Fix compatibility with WP Super Cache

Following up on [this user report](https://wordpress.org/support/topic/labelprintpdf-is-only-ascii/), I discovered that printing labels doesn't work with a plugin (WP Super Cache) that millions of sites use. That's bad.

The problem is that WP Super Cache messes with the response headers, they are converted to `text/html`, so when trying to render a PDF, that header means it will be rendered as text, thus rendering binary garbage. `WP Super Cache` [has a PR](https://github.com/Automattic/wp-super-cache/pull/121) to address this (I think) but it's been in progress for months.

The real fix is on https://github.com/Automattic/woocommerce-services/commit/19a50ad185849f2656e37eb49af656c42fe6c99f but I decided to do a little refactor of our REST classes to make it easier. The 3 commits are independent to make review easier.

Also, fixed #438 while I was there.

To test:
* Install and enable `WP Super Cache`
* Enable the `WP Super Cache` option `Caching on (recommended)`
* Go through any label flow that involves a PDF. Either purchasing (and you should see the Preview, and the real label if you go through the purchase) or reprinting a label.
* On `master`, you should get a garbage text file. On this PR, you should see the PDFs correctly.

Note: Notify the user (ticket #543126) when this is released.
